### ISSUE TITLE
ressol hang on non-prime modulus

### ISSUE THREAD

**Issue by guidovranken** [randombit/botan#2476]
I know `ressol` normally takes a prime, but I wanted to test what would happen if you run it with random numbers.

OSS-Fuzz found a number pair that leads to a hang: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=27333

The numbers (decimal):

```cpp
36201062682172739133895248774702438971945580133434799509985105960684260785324273354862758698920426637444639195349626

7804371375789980578453993074482915734542659201646310600434507062475783156891915333150829678341466565780783986206336267453050408740994685888001
```

I've disabled ressol testing now in Cryptofuzz. This is just to notify you of this test case. Feel free to close this issue without addressing it.

**Comment by randombit** [comment]
Thanks. We've had issues here in the past. The only place it is used is EC point decompression - this does open up a possible DoS if we accept an arbitrary (non-prime `p`) group then try to decompress a point on this curve. So when EC groups are decoded we do a Baillie–PSW test on p and the order, to prevent this. But if ressol itself avoids hanging that is certainly better...

[If nothing else, for safety it may be better to move the primality test into ressol itself]

Out of curiosity, how do other implementations react to non-primes for modular inversion?

**Comment by randombit** [comment]
Interesting! The hang is actually while trying to find an arbitrary quadratic non-residue modulo `p`, which should succeed even with a non-prime. There seems to be a bug in our Jacobi symbol calculation. [Might be worth adding that to cryptofuzz?]

With a fixed Jacobi symbol implementation, your test cases returns -1 (no inverse) quickly.

**Comment by randombit** [comment]
Even more interesting, on further investigation, I don't think this is a bug in the Jacobi symbol calculation. Instead it turns out to be that it's very hard to find a non-residue modulo this integer, and we loop "forever" trying to generate one.

My "fixed" Jacobi symbol code actually was buggy, but in a way that happened to pass the minimal tests for the Jacobi symbol calculation. Then, the incorrect code returned -1 incorrectly, and then the main ressol algorithm proceeded to fail.

I ran this loop in sage

```sage
p = 7804371375789980578453993074482915734542659201646310600434507062475783156891915333150829678341466565780783986206336267453050408740994685888001
z = 1
while kronecker(z, p) == 1:
     if z % 100000 == 0:
         print(z)
     z += 1

print(z)
```

I killed it after testing the first 400 million integers. I then tried a loop with random integers, which also has not found any (!) after 10 minutes.

For a prime, the expected number of tests before finding a non-residue is 2. I imagine the expected number is higher for composites, in a way that depends on the factorization, though I'm not sure about the details.

It looks like OpenSSL just bails out after ~80 iterations, which makes sense since that is very unlikely to occur with a prime. We'll probably have to adopt similar logic.
